src/Bbt.Maybe/Maybe.cs

```diff
@@ -0,0 +1,243 @@
+        /// <param name="aMaybeParameterName">The maybe reference used in error message.</param>
+        /// <param name="aAdditionalMessage">Additional error message.</param>
+        /// <returns>The value.</returns>
+        public T ThrowExceptionIfNone(
+            string aMaybeParameterName,
+            string aAdditionalMessage = "")
+        {
+            if (!this.HasValue)
+            {
+                var lMsg = $"({aMaybeParameterName}) may not be null."
+                    .ToString(CultureInfo.InvariantCulture);
+                if (!string.IsNullOrEmpty(aAdditionalMessage))
+                {
+                    lMsg = lMsg + " " + aAdditionalMessage;
+                }
+
+                throw new InvalidOperationException(lMsg);
+            }
+
+            return this.mValue;
+        }
+
+        /// <summary>
+        /// See <see cref="object.Equals(object)"/>.
+        /// </summary>
+        /// <param name="aObj">The object to compare.</param>
+        /// <returns>True if equal, false otherwise.</returns>
+        public override bool Equals(object aObj)
+        {
+            if (!(aObj is Maybe<T>))
+            {
+                return false;
+            }
+
+            return this.Equals((Maybe<T>)aObj);
+        }
+
+        /// <summary>
+        /// See <see cref="IEquatable{T}.Equals(T)"/>.
+        /// </summary>
+        /// <param name="aOther">The maybe to compare.</param>
+        /// <returns>True if equal, false otherwise.</returns>
+        public bool Equals(Maybe<T> aOther)
+        {
+            return this == aOther;
+        }
+
+        /// <summary>
+        /// See <see cref="object.GetHashCode()"/>.
+        /// </summary>
+        /// <returns>The hash code.</returns>
+        public override int GetHashCode()
+        {
+            if (this.HasValue)
+            {
+                this.mValue.GetHashCode();
+            }
+
+            return base.GetHashCode();
+        }
+
+        /// <summary>
+        /// See <see cref="ISerializable.GetObjectData"/>.
+        /// </summary>
+        /// <param name="aInfo">The serialization info.</param>
+        /// <param name="aContext">The streaming context.</param>
+        public void GetObjectData(SerializationInfo aInfo, StreamingContext aContext)
+        {
+            if (aInfo == null)
+            {
+                throw new ArgumentNullException(nameof(aInfo));
+            }
+
+            aInfo.AddValue(nameof(this.mValue), this.mValue, typeof(T));
+        }
+
+        /// <summary>
+        /// See <see cref="object.ToString"/>.
+        /// </summary>
+        /// <returns>The string representation.</returns>
+        public override string ToString()
+        {
+            if (this.HasValue)
+            {
+                return this.mValue.ToString();
+            }
+
+            return string.Empty;
+        }
+    }
+
+    /// <summary>
+    /// Factory methods.
+    /// </summary>
+    public static class Maybe
```

Comment:
I would put this class in a separate file, especially since it is not only related to `Maybe<T> where T : class` but also `Maybe<T> where T : struct`